Graphene Evaluation

gtalarico opened the pull request:
So far seems to work well - would be so happy to strip out all those views and serializers.
~Need to test some more complex relationships and queries see how this flexes.~


### Todo
- [X] Test Mutations/Write
- [X] Test Filter by multiple hashtags (eg. `companies/?hashtags=web,api` -> `{ companies (hashtags: ["a", "b"] }`)
- [ ] Test Pagination



...

![image](https://user-images.githubusercontent.com/9513968/94336116-5df88700-ff95-11ea-9aa4-4f7e7a2163d8.png)


PS: GraphiQl working 👍 

![image](https://user-images.githubusercontent.com/9513968/94902584-46b21180-044d-11eb-891f-53fe8ad3038f.png)


--- 

### Mutations

**Mutations**
I evaluated graphene's mutations and DRF SerializerMutation.
I think SerializerMutation creates too tight a coupling between model and serialization.
The regular mutation seems adequate enough.

**Auth/Middleware**
While we don't have a full strategy for auth and permissions, the attached middleware examples shows a way to implement something similar to our most common use case (POST/mutations require auth - clap, comment, create, etc).

TODO: not sure where how permission like `isEditor` would be enforced



![image](https://user-images.githubusercontent.com/9513968/94903084-17e86b00-044e-11eb-9a28-24828aa3f05e.png)

danrocha commented:
Does Django's plugin provide a GraphiQL interface to play with the API? 
If so, it is usually more useful than Postman for that... especially if it has an Explorer. Also, it comes with Documentation built in.

Otherwise, you can also try to play with it here https://graphiql-online.com/

danrocha reviewed (approved):
Oh, I see the GraphiQL = true there... nice.

gtalarico commented:
yep! didn't test the UI interface but should be working!
FYI, what's left to do here is to try implementing some of the mutations and see how easily we can replace the existing views.

I am also curios, on the front end, would you suggest just using regular fetch requests with gql payload? or use a client like apollo?

danrocha commented:
Apollo is great, integrated well with Vue, and has its own cache, checks for repeated queries, integrates well with vue, etc etc. etc. there are many more features we'll prob never use though...

However, going in the "simple fetch" direction, there is another interesting library we can check: [Villus](https://github.com/logaretm/villus). I love its simplicity and tried using it a while ago, but it was too early in their development and a bit too buggy for production. They are approaching v1.0 soon though and getting commits every day.

gtalarico commented on api/gql_schema.py:
```diff
@@ -76,6 +79,43 @@ class Meta:
         fields = "__all__"
 
 
+class CreatePostWithSerializer(SerializerMutation):
```
Did not like `SerializerMutation` - too much magic, hard to understand and modify behaviour

gtalarico commented on api/gql_schema.py:
```diff
@@ -76,6 +79,43 @@ class Meta:
+        return kwargs
+
+
+class CreatePost(graphene.Mutation):
```
this seems simple enough and would be really easy to migrate from our current viewsets. 
The only thing is DRF model serializers have better built validation that might need to happen manually here, maybe pre-service calls

gtalarico commented:
* Refactored into sep. files
* Added all/almost all required Queries
* Added a few query filters/search

```
query {
    commentsByThreadId (threadId: 1) {
        text
        clappers {
            name
        }
        profile {
            name
        }
    }
    commentsByParentId (parentId: 1) {
        text
    }
    companies (hashtagNames: "housing") {
        name 
        description
        clapCount
        threadSize
        articles {
            url
        }
        clappers {
            name
        }
        thread { 
            id
        }
        hashtags {
            slug
        }
    }
    posts (hashtagNames: ["social"]) {
        title,
        hashtags {
            slug
        }
    }  
    companyBySlug (slug: "abodu")  {
    name 
    description
    twitterHandle
    articles {
        url
    }
    }
}
```

gtalarico commented:
Looks like django graphene can handle some of the filtering automagically, could simplify things a bit:
https://github.com/graphql-python/graphene-django/blob/master/examples/cookbook/cookbook/ingredients/schema.py#L31

Also should try other mutation base classes instead of creating them manually:
https://docs.graphene-python.org/projects/django/en/latest/mutations/
